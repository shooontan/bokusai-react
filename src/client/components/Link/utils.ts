import { matchPath } from 'react-router-dom';
import { LoadableComponent } from '@loadable/component';
import { CustomRouteConfig } from '~/types/app.type';

export const isElement = (target: unknown): target is Element => {
  return target instanceof Element;
};

export const isLoadableComponent = (
  component: CustomRouteConfig['component']
): component is LoadableComponent<unknown> => {
  if (Object.getOwnPropertyNames(component).includes('preload')) {
    return true;
  }
  return false;
};

export const findRoute = (
  path: string,
  routes: CustomRouteConfig[]
): CustomRouteConfig | undefined => {
  const matchingRoute = routes.find(route =>
    matchPath(path, {
      ...route,
    })
  );
  if (matchingRoute && matchingRoute.routes) {
    return findRoute(path, matchingRoute.routes);
  }
  return matchingRoute;
};
