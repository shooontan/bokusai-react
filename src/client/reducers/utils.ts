export type Unpack<T> = T extends { [K in keyof T]: infer U } ? U : never;

export type ReturnTypes<T> = {
  [K in keyof T]: T[K] extends (...args: any[]) => any ? ReturnType<T[K]> : any;
};

export type ActionTypesCreator<T> = Unpack<ReturnTypes<T>>;
