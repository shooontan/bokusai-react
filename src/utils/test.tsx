import * as React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, RenderOptions, RenderResult } from '@testing-library/react';

const Wrapper: React.FC = ({ children }) => {
  return <BrowserRouter>{children}</BrowserRouter>;
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
