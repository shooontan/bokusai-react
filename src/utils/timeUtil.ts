/**
 * sleep
 */
export const sleep = (timeout: number) =>
  new Promise(resolve => {
    setTimeout(resolve, timeout);
  });
