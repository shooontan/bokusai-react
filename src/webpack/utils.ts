import { resolve } from 'path';

export const isDev = process.env.NODE_ENV !== 'production';

export const pkgdir = resolve(__dirname, '..', '..');

export const getTarget = (target: 'web' | 'node') => target;
