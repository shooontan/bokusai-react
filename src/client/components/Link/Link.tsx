import * as React from 'react';
import { NavLink, NavLinkProps } from 'react-router-dom';
import { routes, NotFoundPage } from '~/app/app';
import { isElement, isLoadableComponent, findRoute } from './utils';

export type CustomLinkProps = NavLinkProps;

export default class CustomLink extends React.Component<NavLinkProps> {
  private linkRef = React.createRef<NavLink>();
  private observer?: IntersectionObserver;
  private isIntersected: boolean = false;

  state = {
    preloaded: false,
  };

  componentDidMount() {
    this.observer = new IntersectionObserver(
      ([entry]) => {
        const ratio = entry.intersectionRatio;
        const next = ratio > 0;
        if (!this.isIntersected && next) {
          this.onEnter();
        }
        this.isIntersected = next;
      },
      {
        root: null,
        rootMargin: '0px',
        threshold: [0],
      }
    );

    if (isElement(this.linkRef.current)) {
      this.observer.observe(this.linkRef.current);
    }
  }

  componentWillUnmount() {
    if (this.observer && isElement(this.linkRef.current)) {
      this.observer.unobserve(this.linkRef.current);
    }
  }

  onEnter = () => {
    if (!this.state.preloaded) {
      this.preloadRouteComponent(this.props.to);
    }
  };

  preloadRouteComponent(to: NavLinkProps['to']) {
    const path = typeof to === 'string' ? to : '/';
    const matchingRoute = findRoute(path, routes);
    if (matchingRoute && isLoadableComponent(matchingRoute.component)) {
      matchingRoute.component.preload();
    } else {
      NotFoundPage.preload();
    }
    this.setState({ preloaded: true });
  }

  handleMouseEnter = () => {
    if (!this.state.preloaded) {
      this.preloadRouteComponent(this.props.to);
    }
  };

  render() {
    return (
      <NavLink
        ref={this.linkRef}
        onMouseEnter={this.handleMouseEnter}
        {...this.props}
      />
    );
  }
}
