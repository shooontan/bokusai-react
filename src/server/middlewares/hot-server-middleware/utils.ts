import path from 'path';
import Module from 'module';
import { MultiCompiler, Compiler, Stats, compilation } from 'webpack';

type CustomCompilation = compilation.Compilation & {
  name?: string;
};

type CunstomStats = Stats & {
  compilation: CustomCompilation;
};

export type MultiStats = {
  stats: CunstomStats[];
  hash: string;
};

export const isMultiCompiler = (
  compiler: Compiler | MultiCompiler
): compiler is MultiCompiler => {
  return compiler && Object.getOwnPropertyNames(compiler).includes('compilers');
};

export const getCompiler = (
  multiCompiler: MultiCompiler,
  name: 'server' | 'client'
) => {
  return multiCompiler.compilers.filter(compiler => compiler.name === name);
};

export const getStats = (multiStats: MultiStats, name: 'server' | 'client') => {
  return multiStats.stats.filter(stats => stats.compilation.name === name);
};

export const getFilename = (
  serverStats: Stats,
  outputPath: string,
  chunkName: string
) => {
  const { assetsByChunkName } = serverStats.toJson();
  const filename = (assetsByChunkName && assetsByChunkName[chunkName]) || '';
  return path.join(
    outputPath,
    Array.isArray(filename)
      ? filename.find(asset => /\.js$/.test(asset)) || ''
      : filename
  );
};

export const requireFromString = (code: string, filename: string) => {
  const parent = module.parent || undefined;
  const m = new Module(filename, parent);
  m.filename = filename;
  // @ts-ignore
  const paths: string[] = Module._nodeModulePaths(path.dirname(filename));
  m.paths = paths;
  // @ts-ignore
  m._compile(code, filename);
  return m.exports;
};
