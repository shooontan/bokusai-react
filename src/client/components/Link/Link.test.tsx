import * as React from 'react';
import { render } from '~/utils/test';
import Link from './Link';

describe('Link', () => {
  test(`to '/'`, () => {
    const { container } = render(<Link to="/" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test(`to '/about'`, () => {
    const { container } = render(<Link to="/about" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
