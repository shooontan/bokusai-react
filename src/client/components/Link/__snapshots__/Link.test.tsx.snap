// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link to '/' 1`] = `
<a
  aria-current="page"
  class="active"
  href="/"
/>
`;

exports[`Link to '/about' 1`] = `
<a
  href="/about"
/>
`;
